packages/builtin/array.pony

```diff
@@ -1,6 +1,139 @@
 class Array[A] is Seq[A]
   """
   Contiguous, resizable memory to store elements of type A.
+
+  # Usage
+  Creating an array of Strings, single line.
+  ```pony
+    let array: Array[String] = ["dog"; "cat"; "wombat"]
+  ```
+
+  Semicolon is an expression, so this is a multiline equivalent.
+   ```pony
+    let array: Array[String] = [
+      "dog"
+      "cat"
+      "wombat"
+    ]
+  ```
+
+  Accessing elements can be done via the `apply(i: USize): this->A ?` method.
+  The index provided might be out of bounds so `apply` is partial and has to be
+  called within a try-catch block or inside another partial method.
+  ```pony
+    fun second_element_is_wombat(array: Array[String]): Bool =>
+      try
+        // first element's index is 0, so second is 1
+        return array(1)? == "wombat"
+      else
+        return false
+      end
+
+      // the second element is indeed a wombat
+      second_element_is_wombat(["star"; "wombat"]) == true
+  ```
+
+  # Memory allocation
+  Array allocates contiguous memory. It always allocates at least enough memory
+  space to hold all of it's elements. Space is the number of elements the array
+  can hold without allocating more memory. The exact amout of memory needed for
+  one unit of space depends on the type of elements the array can hold.
+
+  When creating an array or adding more elements the Array will allocate
+  a next power of two of space, but no less than 8 of space. Empty array
+  is an exception and can be of 0 space.
+
+  Here's a few examples of the space allocated when initialising a table with
+  various number of elements:
+
+  | size | space |
+  |------|-------|
+  | 0    | 0     |
+  | 1    | 8     |
+  | 8    | 8     |
+  | 9    | 16    |
+  | 16   | 16    |
+  | 17   | 32    |
+
+  The space allocated can change when elemnts are added or removed from an
+  array.
+  ```pony
+      // create an array of 8 wombats
+      let wombats = Array[String].init("wombat", 8)
+      wombats.size() == 8
+      wombats.space() == 8
+
+      // insert a cat into the wombats
+      // all elements begining from index 4 will be moved one index to the right
+      try
+        wombats.insert(4, "cat")?
+        wombats.size() == 9
+        wombats.space() == 16
+      end
+
+      // delete fourth element - the cat
+      // all elements on the right will be moved one position to the left
+      try
+        wombats.delete(3)?
+      end
+      // the space of the array won't change, unless the Garbage Collector runs
+      wombats.space() == 16
+      // we can tell the GC to try and reclaim unused space
+      wombats.compact()
+      // this may or not happen
+      wombats.space() == 8 or wombats.space() == 16
+  ```
+
+  It's possible to create an empty array with at some memory allocated.
+  ```pony
+      let empty_array = Array[String].init(9)
+      empty_array.size() == 0
+      empty_array.space() == 16
+  ```
+
+  # More examples
```

Comment:
Most of those examples should be placed in the respective methods they describe.
The class documentation here should focus on an overview of what is available in this class and describe semantics, behavior and usage of this class in general.

Thus is would remove this section here and move some of the content.